Big documentation update

rickecon opened the pull request:
This PR adds UBI directly to `OG-USA`. It also updates the documentation to reflect the default specification of the model. In particular, it adds `r_hh` to the documentation and includes the partially open economy as the default specification. The follow checklist gives the sections that I have gone through that I think are done.
- [x] Model Overview
- [x] Households
- [x] FIrms
- [x] Government
- [x] Market Clearing
- [x] Open Economy Options
- [x] Stationarization
- [x] Equilibrium
- [x] Exogenous Parameters
- [x] Calibration of Macroeconomic Parameters
- [x] Demographics
- [x] Lifetime Earnings Profiles
- [x] Tax Functions
- [x] Bequest Process Calibration
- [x] Universal Basic Income

rickecon commented:
My last commit finished a big update for all the documentation up through the steady-state equilibrium section of `equilibrium.md`. I re-did the description of the steady-state solution algorithm to match our default specification. I am going to leave the steady-state results section as a TODO for us to update later. To finish the equilibrium section of the documentation, I just need to go through the transition path description and make sure it is updated.

rickecon commented:
I ran out of gas on updating `equilibrium.md` in the documentation. I put `TODO`'s in the places that still need to be finished. The steady-state section is done, except for updated results. The non-steady-state solution algorithm is not done. I did do the first paragraphs of text, but I need to finish the algorithm steps. And the non-steady-state results section needs to be updated.

rickecon commented:
@jdebacker . I want to take out the code changes in the four files `SS.py`, `default_parameters.json`, `parameters.py`, and `tax.py`, and submit this PR as a documentation update. Then I will try to get the UBI code updated this weekend in a separate PR. I think there is some demand for the updated documentation (e.g., careful descriptions of derivations of `r_hh`, `r_t`, and `r_gov`.

rickecon commented:
@jdebacker. This PR is ready for your review (conditional on all tests passing). I would recommend the following process for reviewing this PR.
1. Create your own branch copy of this branch.
```
git checkout -b rickecon-ubi master
git pull https://github.com/rickecon/OG-USA.git ubi
```
2. Activate the `ogusa-dev` conda environment and compile the updated Jupyter Book documentation on your own machine.
```
jb build ./docs/book
```
3. Give the documentation a read from the compiled HTML, guided by the file diff in GitHub.

Here are the main changes.
* Updated two references in the `OGUSA_references.bib` file
* Updated copyright year to 2021 in `_config.yml`
* Moved the location of the `open_economy` chapter within the Theory section and added two more chapters (`UBI` and `matching_lwi`) in the Calibration section in the `_toc.yml` file.
* Added UBI directly to the model theory, exactly as it will be added in the code.
* Set the baseline specification of the model to large open economy
* Specified `r_hh`, `r_gov`, and `r_t` throughout the theory.
* Updated the steady-state solution method description for the baseline case.

jdebacker commented on docs/book/content/contributing/contributor_guide.md:
````diff
@@ -72,19 +51,19 @@ installation that came as part of your computer's operating system.
+      OG-USA$ conda activate ogusa-dev
     ```
-    If you are working on Windows, use the following from the command line::
+    If you are working on Windows, use the following from the command line:
````
These separate instructions for Windows can be deleted. Now all OS's use `conda activate`.

rickecon replied:
Removed separate Windows OS instruction. Thanks @jdebacker

jdebacker commented on docs/book/content/contributing/contributor_guide.md:
```diff
@@ -95,10 +74,20 @@ section introduces some standard Git practices and guides you through
+(Sec_GitHubIssue)=
+### Submitting a GitHub Issue
+
+GitHub "issues" are an excellent way to ask questions, include code examples, and tag specific GitHub users. [TODO: Add more on GitHub issues]
```
Unless it is specific to OG-USA, it's probably better to reference the `Git-Tutorial` repo so that we aren't duplicating any more than necessary across PSL projects.

rickecon replied:
@jdebacker . I fully agree with this comment. But the following two considerations make me think we should leave this section as it stands for now.
1. I didn't really add much to this section, I just reformatted most of the points (with the exception of the "Submitting a GitHub Issue" section). We really need to update everything here from the "Setup Git" section and below. We should be referencing all of that to the [Git-Tutorial](https://pslmodels.github.io/Git-Tutorial/) documentation.
2. The [Git-Tutorial](https://pslmodels.github.io/Git-Tutorial/) documentation is not up to date. I think as soon as we have what we want from this section in the Git-Tutorial, then we rewrite this section of the OG-USA contributor guide to point to the Git-Tutorial.

So for now, I recommend that we leave this as it is.

rickecon replied:
@jdebacker. Per your comment below, I have removed the "TODO" comment about adding more GitHub issues.

jdebacker commented:
@rickecon Thanks for this much needed refresh to the documentation and the addition of the UBI system!  Besides my comments on specific lines, I have the following:

- [ ] Most of equations in stationarization chapter are not rendering well - though they do when printed to pdf (at least the ones on separate lines).  This could just a Jupyter-Book issue.
- [ ] In “Steady-state solution method: default specification”: "...and in which baseline government spending is not held constant until the closure rule (baseline_spending = False)".  This description could be more clear.  The baseline spending option only matters for reform runs (where it holds the amount of `TR` and `G` at the levels from the baseline).
- [ ]  Same as above for non-steady state solution
- [ ] Non-steady state results still TODO, but will need to put open econ in there and interest rate wedge
- [ ] - First sentence of “Calculating UBI” section: “UBI household transfers”
- [ ] UBI parameter names - do we want to make these consistent with Tax-Calculator where possible?  Do we want to be able to read them in from a Tax-Calculator reform dictionary if they are specified there? This second question related to the first but is not relevant for documentation (it'd be handled in `parameters.py`, I think).
- [ ] In the "UBI specification not adjusted for economic growth" section, there needs to be a space between "nonstationary" and "values" in first sentence of last paragraph.  This whole sentence should be looked at, it could be compilation issues, but it seemed a bit jumbled.

jdebacker reviewed (commented):
@rickecon I am ok with the section, but think the "TODO" can be dropped - those extra things are/should be in Git-Tutorial, unless they happen to be project specific.

rickecon commented:
> Most of equations in stationarization chapter are not rendering well - though they do when printed to pdf (at least the ones on separate lines). This could just a Jupyter-Book issue.

All of the equations in the "Stationarization" chapter are rendering well on my machine in HTML (using Chrome browser on MacBook Pro, OS Big Sur v 11.2.2). I have found that equation rendering in the HTML version of Jupyter Book documentation is usually a function of page magnification and window size. Try making your window larger and the magnification smaller. My rendering of multi-equation lines looks like the following:

<img width="848" alt="Screen Shot 2021-03-09 at 10 52 52 AM" src="https://user-images.githubusercontent.com/4317690/110515419-9f08d180-80c5-11eb-9b22-5fe86ef5e209.png">

jdebacker commented:
@rickecon Got it - equations look great in Chrome.  I was using Safari, where they didn't render as well.  Thanks for pointing that out - no idea.

rickecon commented:
@jdebacker said:
> In “Steady-state solution method: default specification”: "...and in which baseline government spending is not held constant until the closure rule (baseline_spending = False)". This description could be more clear. The baseline spending option only matters for reform runs (where it holds the amount of TR and G at the levels from the baseline).

I updated this wording in the "Steady-state solution method: default specification" section to read:
> ...and in which baseline government spending `G` and transfers `TR` are not held at baseline levels (`baseline_spending = False`).

I think this is clearer.

rickecon commented:
@jdebacker said in box check list above:
>  Same as above for non-steady state solution
>
> Non-steady state results still TODO, but will need to put open econ in there and interest rate wedge

The entire section "Stationary non-steady-state solution method: default specification" needs to be redone following the pattern of the steady-state solution method section. It just takes a lot of time, so I left a TODO there. I recommend that we leave this to a future documentation PR.

On the topic of TODO messages in the text, you say:
>Here and throughout, I would recommend using GH Issues to highlight work to be done rather than inline "TODO" statements. These work better in code than in the compiled documentation. GH Issues also more easily raise awareness for other contributors.

I  definitely think we should have GH issues for every place there is a TODO. But I think the TODO in the text of the documentation is nice because it shows the reader that we are working on that and are aware that something is missing.

rickecon commented on docs/book/content/calibration/bequests.md:
```diff
@@ -1,6 +1,8 @@
 (Chap_Beq)=
 # Bequest Process Calibration
 
+[TODO: This chapter needs to be finished.]
```
@jdebacker. I definitely think we should have GH issues for every place there is a TODO. But I think the TODO in the text of the documentation is nice because it shows the reader that we are working on that and are aware that something is missing.

rickecon commented on docs/book/content/theory/market_clearing.md:
```diff
@@ -1,48 +1,128 @@
 # Market Clearing
 
-Three markets must clear in `OG-USA`---the labor market, the capital market, and the goods market. By Walras' Law, we only need to use two of those market clearing conditions because the third one is redundant. In the model, we choose to use the labor market clearing condition and the capital market clearing condition, and to ignore the goods market clearing condition. But we present all three ma ...
+Four markets must clear in `OG-USA`---the labor market, the private capital market, the government bonds market, and the goods market. By Walras' Law, we only need to use three of those market clearing conditions because the fourth one is redundant. In the model, we choose to use the labor, private capital, and government bonds market clearing conditions and to ignore the goods market clearing co ...
```
Done. I just fixed this.

rickecon commented on docs/book/content/theory/market_clearing.md:
```diff
@@ -1,48 +1,128 @@
+  (SecMarkClrMktClr_CapGen)=
+  ### Capital markets generalities
+
+  Before describing the government bond market and private capital market clearing conditions, respectively, we define some general capital market conditions relative to both markets. Both the government bond market and private capital market are characterized by interest rates that differ exogenously. As described in {eq}`EqUnbalGBC_rate_wedge`, the interest rate at which the government repays d ...
```
Done. Fixed this.

rickecon commented on docs/book/content/intro/intro.md:
```diff
@@ -1,15 +1,30 @@
 # OG-USA
 
-OG-USA is an overlapping-generations (OG) model of the economy of the United States (USA) that allows for dynamic general equilibrium analysis of federal tax policy. The model output focuses changes in macroeconomic aggregates (GDP, investment, consumption), wages, interest rates, and the stream of tax revenues over time. Documentation of the model theory--its output, and solution method--is avai ...
+`OG-USA` is an overlapping-generations (OG) model of the economy of the United States (USA) that allows for dynamic general equilibrium analysis of federal tax policy. The model output focuses changes in macroeconomic aggregates (GDP, investment, consumption), wages, interest rates, and the stream of tax revenues over time. This documentation contains the following major sections, which are regul ...
```
@jdebacker . Changed to "federal fiscal policy". Thx.

rickecon commented on docs/book/content/theory/open_economy.md:
```diff
@@ -1,137 +1,46 @@
 
-(SecOpenRC)=
-#### Resource Constraint
+Note that in our partially open economy specification, the world interest rate $r^*_t$ is necessary for determining equilibrium because the foreign supply of private capital $K^f_t$ depends on a concept of excess demand that is based on domestic demand at the world interest rate {eq}`EqMarkClr_ExDemK`. It is also worth noting that our partially open economy specification is not a multi-country mo ...
```
@jdebacker. I added the following phrase at the end of this paragraph:
> as influenced by the supply of foreign private capital `zeta_K` and by foreign purchases of new issues of government bonds `zeta_D`.

rickecon commented:
@jdebacker said:
> First sentence of “Calculating UBI” section: “UBI household transfers”
>
> In the "UBI specification not adjusted for economic growth" section, there needs to be a space between "nonstationary" and "values" in first sentence of last paragraph. This whole sentence should be looked at, it could be compilation issues, but it seemed a bit jumbled.

I fixed the typo in the first section of the "Calculating UBI" section, and I rewrote the last paragraph of the "UBI specification not adjusted for economic growth" section to be two paragraphs with a new equation in between, plus a new footnote. They are the following:

<img width="836" alt="Screen Shot 2021-03-09 at 3 09 19 PM" src="https://user-images.githubusercontent.com/4317690/110544989-bce82d80-80e9-11eb-8b1f-70114dc61d0e.png">

rickecon commented:
@jdebacker said:
> UBI parameter names - do we want to make these consistent with Tax-Calculator where possible? Do we want to be able to read them in from a Tax-Calculator reform dictionary if they are specified there? This second question related to the first but is not relevant for documentation (it'd be handled in `parameters.py`, I think).

I don't know about this. Conceptually it is always nice to have OG-USA read parameters from Tax-Calculator or any other model. But:
1.  OG-USA's UBI is a mutually exclusive substitute for whatever capability Tax-Calculator has to handle UBI. So I am trying to think of the use case for bringing those parameters in from Tax-Calculator.
2. We can always make a crosswalk from Tax-Calculator to OG-USA in `parameters.py`.
3. The Tax-Calculator UBI parameters are `UBI_u18`, `UBI_1820`, `UBI_21`, `UBI_ecrt`. These are a little different from the current OG-USA parameters in PR #715.

My inclination is to keep these separate for now. But it would be straightforward to harmonize them in the future. @MaxGhenis signed off on the OG-USA parameters when I set them up.

rickecon commented:
@jdebacker . I think I have addressed all your questions in the discussion above. Let me know what else you would like me to change for this to be merged. Thanks for the long review.

MaxGhenis commented:
Yeah I don't think we can avoid the confusion here on parameters, but separate names from the taxcalc parameters seems like the right call IMO.

Will the UBI go in the `OG-USA Parameters...Fiscal Policy Parameters` section of Compute Studio, rather than `Tax-Calculator Parameters`?

rickecon commented:
@MaxGhenis said:
> Will the UBI go in the OG-USA Parameters...Fiscal Policy Parameters section of Compute Studio, rather than Tax-Calculator Parameters?

I haven't crossed that bridge yet. But I don't think it makes sense to put them in the Tax-Calculator parameters since they will not be Tax-Calculator parameters.

jdebacker commented:
@rickecon Said:

> I don't know about this. Conceptually it is always nice to have OG-USA read parameters from Tax-Calculator or any other model. But:
> OG-USA's UBI is a mutually exclusive substitute for whatever capability Tax-Calculator has to handle UBI. So I am trying to think of the use case for bringing those parameters in from Tax-Calculator.

The use case would be the same as with taxes: one wants to get a detailed static revenue estimate and distributional analysis and so simulates a UBI in Tax-Calculator.  Then one further wants to find the economic impacts of this plan and so runs the UBI program through OG-USA.  By mapping parameters from Tax-Calculator into OG-USA, the user only needs to specify the parameters one time and it ensures consistency.

> We can always make a crosswalk from Tax-Calculator to OG-USA in parameters.py.

This is a good point  - we could do that.  But we are starting from scratch now, so we can avoid having to do work later if we want.

> The Tax-Calculator UBI parameters are UBI_u18, UBI_1820, UBI_21, UBI_ecrt. These are a little different from the current OG-USA parameters in PR #715.

Yes, but we could (in principle) add one more parameter for UBI for 18-20 year olds to OG-USA and a UBI for 65+ to Tax-Calculator.  Seems like the latter would be a good addition to Tax-Calculator - kind of surprised it's not there.  The UBI for 18-20 year olds in OG-USA could be tough to calibrate since they aren't economically active in the model at that age, but in the data there maybe many of that age in their own households (not dependents of their parents).

cc @MaxGhenis 

> My inclination is to keep these separate for now. But it would be straightforward to harmonize them in the future. @MaxGhenis signed off on the OG-USA parameters when I set them up.

MaxGhenis commented:
Adding UBI_65 is blocked by a taxdata bug, see https://github.com/PSLmodels/Tax-Calculator/issues/2469.

I think taxcalc split out 18-20-year-olds because AEI scholar Charles Murray's UBI plan is 21+, but all others I've seen have been 18+. If it complicates things in OG-USA it could be another reason to consider dropping it - I doubt it's been used much if at all.

The 0-17, 18-64, 65+ buckets are the right ones to start with IMO, and if any would be added next, it'd probably be splitting the child component into 0-5 and 6-17, to match US child allowance proposals and address differences in childcare needs by age.

rickecon commented:
@jdebacker @MaxGhenis. I like the idea of changing the names of the variables and adding one extra:
* `UBI_dol_u17`: dollar amount of UBI per year for each child (non-head of household and non-spouse of head of household) age 0-17 (under 18)
* `UBI_dol_1820`: dollar amount of UBI per year for each non-head of household and non-spouse of head of household age 18-20
* `UBI_dol_2164`: dollar amount of UBI for each adult ages 21 through 64
* `UBI_dol_65p`: dollar amount of UBI for each adult age 65 and over

These variables are much closer to the Tax-Calculator variable names. But I think it is important to have the "`dol`" descriptor in these OG-USA variable names to distinguish between the dollar unit inputs and the model units that are solved for in the steady-state solution. These variable names are also a little different than the Tax-Calculator names in some other dimensions that I think are better and more descriptive (e.g., "`_u17`" rather than "`_u18`").

rickecon commented:
@jdebacker . I like the `_nom` and `_mod` descriptors for "nominal" and "model" values, respectively. I think the distinction is important in the OG models where inputs are in nominal currency terms and internal values have to be converted to model units. So how do you feel about the following variables?
* `UBI_nom_017`: dollar (nominal) amount of UBI per year for each child (non-head of household and non-spouse of head of household) age 0-17 (under 18)
* `UBI_nom_1820`: dollar (nominal) amount of UBI per year for each non-head of household and non-spouse of head of household age 18-20
* `UBI_nom_2164`: dollar (nominal) amount of UBI for each adult ages 21 through 64
* `UBI_nom_65p`: dollar (nominal) amount of UBI for each adult age 65 and over

rickecon commented:
I just pinned the version of Jupyter Book to v. 0.9.1 in the `environment.yml` file and recompiled the Jupyter Book. Now all the citations in the text are good, the references page compiles, and the citations page compiles--all the required behavior from Issue #717.

Now my question is how to make it work with Jupyter Book version 0.10.0. Maybe we have to add the `citations.bib` file to the `bibtex_bibfiles:` section at the end of `_config.yml`. @jdebacker

jdebacker commented:
> So how do you feel about the following variables?

I am good with those name.  Thanks for thinking about this.

rickecon commented:
I changed the `environment.yml` to be `jupyter-book>=0.9.1`. However, I still can't get the Citations page to compile under the Jupyter Book 0.10.1 or 0.10.0 versions. I tried the following two fixes with v. 0.10.1 in the `_config.yml` file, neither of which worked:

**Fix attempt 1**
```
bibtex_bibfiles:
   - OGUSA_references.bib
   - ../../citations.bib
```

**Fix attempt 2, copying citations.md to same directory as OGUSA_references.bib**
```
bibtex_bibfiles:
   - OGUSA_references.bib
   - citations.bib
```
I think we should move forward with this PR as it is, and work to fix the citations page in a future PR, following the Issue #717 thread.

rickecon commented:
@jdebacker . I have now updated the UBI variable names according to the discussion above in `UBI.md` from the Calibration section of the documentation. I will update those variable names accordingly in PR #715.

rickecon commented:
@jdebacker . I think I am done with all the fixes that will make it into this PR. This is ready for your (maybe) final review.

rickecon commented:
@jdebacker . I don't know what I did, but the Jupyter Book v. 0.10.1 is now compiling the Citations page, with `citations.bib` not in the `_config.yml` file and located in the main `OG-USA/` directory, and with no changes to `citations.md`. OK, now even better to go. Closing Issue #717.

jdebacker commented:
@rickecon This looks great.  Thanks for your hard work on this substantial update!
